org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/dom/parser/XMLScanner.java

```diff
@@ -443,112 +435,161 @@ TokenType internalScan() {
+						return finishToken(offset, TokenType.DTDStartNotation);
 					} else if (stream.advanceIfChars(_EXL, _MIN, _MIN)) { // !-- (for comment)
 						state = ScannerState.WithinComment;
 						return finishToken(offset, TokenType.StartCommentTag);
 					}
 				}
+				if (stream.advanceUntilCharOrNewTag(_RAN)) { // >
+					stream.advanceIfChar(_RAN); // >
+					return finishToken(offset, TokenType.Content);
+				}
+
 			}
 
-			if (stream.advanceUntilChar(_CSB)) { // ]
+			if (stream.advanceIfChar(_CSB)) { // ]
 				state = ScannerState.DTDWithinDoctype;
 				isInsideDTDContent = false;
 				return finishToken(offset, TokenType.DTDEndInternalSubset);
 			}
-			stream.advanceUntilChar(_LAN); // <
+
+			stream.advanceUntilAnyOfChars(_LAN, _CSB); // < || ]
 			return finishToken(offset, TokenType.Content);
+			
+
+		case DTDUnrecognizedParameters:
 
-		case DTDWithinElement:
 			if (stream.skipWhitespace()) {
 				return finishToken(offset, TokenType.Whitespace);
 			}
 
-			if (stream.advanceIfChar(_RAN)) { // >
+			if(stream.advanceIfChar(_RAN)) { // >
 				state = ScannerState.DTDWithinContent;
 				return finishToken(offset, TokenType.DTDEndTag);
 			}
 
-			if (!stream.advanceIfRegExp(Constants.ELEMENT_NAME_REGEX).equals("")) {
-				state = ScannerState.DTDAfterElementName;
-				return finishToken(offset, TokenType.DTDElementDeclName);
+			if(stream.peekChar() == _LAN) { // <
+				state = ScannerState.DTDWithinContent;
+				return internalScan();
+			}
+
+			if(stream.advanceIfChar(_CSB)) { // ]
+				state = ScannerState.DTDWithinDoctype;
+				isDeclCompleted = true;
+				return finishToken(offset, TokenType.DTDEndInternalSubset);
+			}
+
+			if(stream.advanceUntilCharOrNewTag(_RAN)) { // >
+				if(stream.peekChar() == _LAN) { // <
+					state = ScannerState.DTDWithinContent;
+				}
+				return finishToken(offset, TokenType.DTDUnrecognizedParameters);
+			}
+
+			return finishToken(offset, TokenType.Unknown);
+
+		case DTDWithinElement:
+			if (stream.skipWhitespace()) {
+				return finishToken(offset, TokenType.Whitespace);
 			}
 
 			if (stream.advanceIfChar(_RAN)) { // >
 				state = ScannerState.DTDWithinContent;
 				return finishToken(offset, TokenType.DTDEndTag);
 			}
 
-			state = ScannerState.DTDWithinContent;
+			if(isDeclCompleted == true) {
+				state = ScannerState.DTDUnrecognizedParameters;
+				return internalScan();
+			}
+
+			if (!stream.advanceIfRegExp(Constants.ELEMENT_NAME_REGEX).equals("")) {
+				state = ScannerState.DTDElementAfterName;
+				return finishToken(offset, TokenType.DTDElementDeclName);
+			}
+
+			state = ScannerState.DTDUnrecognizedParameters;
 			return internalScan();
 
-		case DTDAfterElementName:
-			nbBraceOpened = 0;
+		case DTDElementAfterName:
 			if (stream.skipWhitespace()) {
 				return finishToken(offset, TokenType.Whitespace);
 			}
 
 			if (stream.advanceIfChar(_ORB)) { // (
-				nbBraceOpened++;
-				state = ScannerState.DTDWithinElementContent;
+				nbBraceOpened = 1;

```

Comment:
nbBraceOpened++; seems safer, always accurate

Reply:
I changed this because at this point it should always equal one. Before it relied on resetting the value back to 0.